// plywood/build/external/sqlExternal.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { Dataset } from '../datatypes/index';
import { Expression, NumberBucketExpression, TimeBucketExpression } from '../expressions/index';
import { External, TotalContainer } from './baseExternal';
function correctResult(result) {
    return Array.isArray(result) && (result.length === 0 || typeof result[0] === 'object');
}
function getSplitInflaters(split) {
    return split.mapSplits(function (label, splitExpression) {
        var simpleInflater = External.getSimpleInflater(splitExpression, label);
        if (simpleInflater)
            return simpleInflater;
        if (splitExpression instanceof TimeBucketExpression) {
            return External.timeRangeInflaterFactory(label, splitExpression.duration, splitExpression.getTimezone());
        }
        if (splitExpression instanceof NumberBucketExpression) {
            return External.numberRangeInflaterFactory(label, splitExpression.size);
        }
        return undefined;
    });
}
function valuePostProcess(data) {
    if (!correctResult(data)) {
        var err = new Error("unexpected result (value)");
        err.result = data;
        throw err;
    }
    return data.length ? data[0][External.VALUE_NAME] : 0;
}
function totalPostProcessFactory(inflaters, zeroTotalApplies) {
    return function (data) {
        if (!correctResult(data)) {
            var err = new Error("unexpected total result");
            err.result = data;
            throw err;
        }
        var datum;
        if (data.length === 1) {
            datum = data[0];
            for (var _i = 0, inflaters_1 = inflaters; _i < inflaters_1.length; _i++) {
                var inflater = inflaters_1[_i];
                inflater(datum, 0, data);
            }
        }
        else if (zeroTotalApplies) {
            datum = External.makeZeroDatum(zeroTotalApplies);
        }
        return new TotalContainer(datum);
    };
}
function postProcessFactory(inflaters, zeroTotalApplies) {
    return function (data) {
        if (!correctResult(data)) {
            var err = new Error("unexpected result");
            err.result = data;
            throw err;
        }
        var n = data.length;
        for (var _i = 0, inflaters_2 = inflaters; _i < inflaters_2.length; _i++) {
            var inflater = inflaters_2[_i];
            for (var i = 0; i < n; i++) {
                inflater(data[i], i, data);
            }
        }
        if (n === 0 && zeroTotalApplies) {
            data = [External.makeZeroDatum(zeroTotalApplies)];
        }
        return new Dataset({ data: data });
    };
}
export var SQLExternal = (function (_super) {
    __extends(SQLExternal, _super);
    function SQLExternal(parameters, dialect) {
        _super.call(this, parameters, dummyObject);
        this.dialect = dialect;
    }
    SQLExternal.prototype.canHandleTotal = function () {
        return true;
    };
    SQLExternal.prototype.canHandleFilter = function (filter) {
        return true;
    };
    SQLExternal.prototype.canHandleSplit = function (split) {
        return true;
    };
    SQLExternal.prototype.canHandleSplitExpression = function (ex) {
        return true;
    };
    SQLExternal.prototype.canHandleApply = function (apply) {
        return true;
    };
    SQLExternal.prototype.canHandleSort = function (sort) {
        return true;
    };
    SQLExternal.prototype.canHandleLimit = function (limit) {
        return true;
    };
    SQLExternal.prototype.canHandleHavingFilter = function (havingFilter) {
        return true;
    };
    SQLExternal.prototype.getQueryAndPostProcess = function () {
        var _a = this, source = _a.source, mode = _a.mode, applies = _a.applies, sort = _a.sort, limit = _a.limit, derivedAttributes = _a.derivedAttributes, dialect = _a.dialect;
        var query = ['SELECT'];
        var postProcess = null;
        var inflaters = [];
        var zeroTotalApplies = null;
        var from = "FROM " + this.dialect.escapeName(source);
        var filter = this.getQueryFilter();
        if (!filter.equals(Expression.TRUE)) {
            from += '\nWHERE ' + filter.getSQL(dialect);
        }
        switch (mode) {
            case 'raw':
                var selectedAttributes = this.getSelectedAttributes();
                selectedAttributes.forEach(function (attribute) {
                    var name = attribute.name, type = attribute.type;
                    switch (type) {
                        case 'BOOLEAN':
                            inflaters.push(External.booleanInflaterFactory(name));
                            break;
                        case 'SET/STRING':
                            inflaters.push(External.setStringInflaterFactory(name));
                            break;
                    }
                });
                query.push(selectedAttributes.map(function (a) {
                    var name = a.name;
                    if (derivedAttributes[name]) {
                        return Expression._.apply(name, derivedAttributes[name]).getSQL(dialect);
                    }
                    else {
                        return dialect.escapeName(name);
                    }
                }).join(', '), from);
                if (sort) {
                    query.push(sort.getSQL(dialect));
                }
                if (limit) {
                    query.push(limit.getSQL(dialect));
                }
                break;
            case 'value':
                query.push(this.toValueApply().getSQL(dialect), from, dialect.constantGroupBy());
                postProcess = valuePostProcess;
                break;
            case 'total':
                zeroTotalApplies = applies;
                query.push(applies.map(function (apply) { return apply.getSQL(dialect); }).join(',\n'), from, dialect.constantGroupBy());
                postProcess = totalPostProcessFactory(inflaters, zeroTotalApplies);
                break;
            case 'split':
                var split = this.getQuerySplit();
                query.push(split.getSelectSQL(dialect)
                    .concat(applies.map(function (apply) { return apply.getSQL(dialect); }))
                    .join(',\n'), from, split.getShortGroupBySQL());
                if (!(this.havingFilter.equals(Expression.TRUE))) {
                    query.push('HAVING ' + this.havingFilter.getSQL(dialect));
                }
                if (sort) {
                    query.push(sort.getSQL(dialect));
                }
                if (limit) {
                    query.push(limit.getSQL(dialect));
                }
                inflaters = getSplitInflaters(split);
                break;
            default:
                throw new Error("can not get query for mode: " + mode);
        }
        return {
            query: query.join('\n'),
            postProcess: postProcess || postProcessFactory(inflaters, zeroTotalApplies)
        };
    };
    SQLExternal.type = 'DATASET';
    return SQLExternal;
}(External));
